CFP-32810: Add hybrid routing mode in Cilium (#73)
cilium/design-cfps: cilium/CFP-32810-hybrid-routing-mode.md

Reviewer: I think that at some point we may want to consider a more formal CRD for this and other related configurations, but... well, I've tried to draft such a CRD in the past and there is a *lot* of possible interactions to consider so I don't really want to open that can of the worms too early. Therefore I think that specifying the subnets as config flags is probably the best option we have today.

```
subnet-topology: 10.0.0.1/24,10.10.0.1/24
```

The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
```
subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24
```

* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
* If source and destination IP belongs to `10.20.0.1/24`, route natively
* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

### Control plane implementation

Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger the eBPF map reconcile.

Here's how the different topologies would be represented in-memory cache for the following config: `10.0.0.1/24,10.10.0.1/24;10.20.0.1/24;2001:0db8:85a3::/64`

| CIDR | Subnet ID |
|------|-----------|
| 10.0.0.1/24 | 1  |
| 10.10.0.1/24 | 1 |
| 10.20.0.1/24 | 2 |

> | 2001:0db8:85a3::/64 | 3 |

Reviewer: Is there an assumption here that not all subnets are mutually routable? That is, a Cilium-managed node may be connected to multiple subnets, but other hosts may not be connected to the same subnets?

Reviewer: ^ I would suggest putting that in the goals of the CFP to state the assumption up front.

#### Notes:

- Subnet ID allocation starts from 1

> - If config changes, replace the entire cache with new Subnet ID calculation instead of trying to reconcile the diff. Predicating it on the assumption that subnet segmentation would be limited (like not 100s of CIDR blocks for a cluster).

Author: Yeah, I thought about this as well. My intention is to make the map updates atomic (maybe spin locks), but I don't want to set it in stone in the CFP itself. I think PR would be a good place to discuss the specific implementation (lock or revision number)? Based on test/performance implication (if any).

> - If config changes, replace the entire cache with new Subnet ID calculation instead of trying to reconcile the diff. Predicating it on the assumption that subnet segmentation would be limited (like not 100s of CIDR blocks for a cluster).

Reviewer: I think this makes sense. Probably worth calling this concern out somewhere in the CFP just to make sure we can come back and review it. I'm fine with not settling on a solution in the CFP before you've had a chance to dig into the options and figure out what works best.

There's two impacts I can think of that relate to this with the subnet ID allocation:
1. If Cilium is running and the subnets are updated, Cilium should retain connectivity for existing connections between connected subnets that remain connected both before and after the subnet configuration change
2. When restarting or upgrading Cilium, Cilium should likewise retain the connectivity.

Worth noting as well that this interface design specifically avoids delegating the subnet ID allocation problem to the user.

> - If config changes, replace the entire cache with new Subnet ID calculation instead of trying to reconcile the diff. Predicating it on the assumption that subnet segmentation would be limited (like not 100s of CIDR blocks for a cluster).

Reviewer: Does it make sense to add a "key question" for this like "How can we minimize traffic impact when changing subnets or upgrading/downgrading Cilium?"

Even if we just state that this question requires investigation before we can agree on an approach, that would be OK. We don't have to resolve this key question during the CFP stage, I just want to make sure it's clearly noted so we don't forget about it :)

> - If config changes, replace the entire cache with new Subnet ID calculation instead of trying to reconcile the diff. Predicating it on the assumption that subnet segmentation would be limited (like not 100s of CIDR blocks for a cluster).

Author: Great idea. I have added our discussions under `Key Questions` that should be evaluated during implementation.

# CFP-32810: Add hybrid routing mode in Cilium

**SIG: SIG-NAME** SIG-Datapath

**Begin Design Discussion:** 2025-06-16

**Cilium Release:** 1.19

> **Authors:** Anubhab Majumdar <anmajumdar@microsoft.com>, Vamsi Kalapala <vakr@microsoft.com>, Krunal Jain <krunaljain@microsoft.com>

Reviewer 2: ```suggestion
**Authors:** Anubhab Majumdar <anmajumdar@microsoft.com>, Vamsi Kalapala <vakr@microsoft.com>, Krunal Jain <krunaljain@microsoft.com>

**Status:**
```

- If no, tunneling happens exactly as it works today 

Above behavior simply optimizes current encapsulation behavior, without introducing major change to how encapsulation routing works/behaves. 

## Impacts

### Impact: Network throughput

Given that today Cilium is encapsulating packets that could be natively routed, removing those scenarios would yield higher throughput. This is especially true in clusters without jumbo frames.

### Impact: Datapath latency

In a highly fragmented network topography, lookup of IP to make routing decision may not be negligible.

## Key Questions

### Connectivity disruption

Things to consider during implementation to minimize traffic disruption:
- how to ensure atomic/revision-based update to the subnet eBPF map, and minimize traffic disruption because of the update
- since subnets can be updated on the fly, it is important to retain connectivity for existing connections between connected subnets that remain connected both before and after the subnet configuration change
- when restarting or upgrading Cilium, datapath should likewise retain the connectivity

> 

Reviewer: Assuming reviewers are good with this:

```suggestion

Resolving this question is deferred to the implementation phase when we expect to learn more about the solution space.

```

Author: Added to the doc.

* Support for both IPV4 and IPV6
* Allow dynamic configuration of CIDRs (add/remove) without agent restart (update w/o agent restart)
* Support for multiple subnets that are not not all mutually routable

## Proposal

### Overview

Add a new option for routing-mode config as follows `routing-mode: hybrid`. Also, add a new field in config to pass information about the CIDRs and connectivity. The CIDR information will persist in pinned eBPF maps so that routing decision can be made in runtime by eBPF programs. The concept is similar to [IP Masquerading](https://docs.cilium.io/en/stable/network/concepts/masquerading/#ebpf-based) and the eBPF table it uses.

### Examples

```
subnet-topology: 10.0.0.1/24
```

If source and destination IP belongs to above CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to known CIDRs/pass to Linux stack).

```
subnet-topology: 10.0.0.1/24,10.10.0.1/24
```

The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
```

> subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24

Reviewer 3: what's the usecase for this case `10.0.0.1/24,10.10.0.1/24;10.20.0.1/24` ? I couldn't understand this

Reviewer 3: does this represent in a cluster, pods belong to`10.0.0.1/24` or `10.10.0.1/24` can't natively route to pods belong to  `10.20.0.1/24` but pods within the range can be natively routed. So this is for hybrid routing within a cluster?

Author: Yes. Comma separated subnets are natively routable, and `;` represents subnets which needs tunneling. This notation will support hybrid routing in a cluster.

Reviewer 3: good to clarify this in CFP

```

* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
* If source and destination IP belongs to `10.20.0.1/24`, route natively
* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

### Control plane implementation

Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger the eBPF map reconcile.

> Here's how the different topologies would be represented in-memory cache for the following config: `10.0.0.1/24,10.10.0.1/24;10.20.0.1/24;2001:0db8:85a3::/64`

Reviewer 3: why configmap and not CRD? Isn't this easy to represent in CRD than configmap?

Author: We discussed this here - https://github.com/cilium/design-cfps/pull/73#discussion_r2199013817 .
Possibility of adding a CRD is a future milestone.

# CFP-32810: Add hybrid routing mode in Cilium

> **SIG: SIG-NAME** SIG-Clustermesh

Author: Agreed. I have cc'd datapath SIG for feedback.

**Begin Design Discussion:** 2025-06-16

> **Cilium Release:** N/A

Reviewer 2: is there a release you want to target?

Author: Preferably 1.18, but I guess it would depend on feedback on this CFP and what the final scope of the work would be.

**Authors:** Anubhab Majumdar <anmajumdar@microsoft.com>, Vamsi Kalapala <vakr@microsoft.com>, Krunal Jain <krunaljain@microsoft.com>

## Summary

> Currently, Cilium supports two modes routing - [encapsulation](https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation) and [native-routing](https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing). When workloads are on cross subnets with no direct connection, Cilium needs to run in encapsulation mode. This is not efficient, as every packet needs to be encapsulated, even when they are on same subnet, thus paying the MTU overhead and reducing network throughput. This CFP proposes adding a third option to allow Cilium dataplane to make the decision at runtime on whether to route the packet natively or encapsulate.

Author: My mistake; I meant encapsulation overhead. Fixed.

## Motivation

Consider any of the following scenarios:

1. Two Kubernetes cluster connected through cluster mesh. The nodes are backed by different subnets in two different VNETs, which are connected through VNET peering. The pods are assigned IP address from overlay network.
2. In a large Kubernetes cluster, pod addresses are assigned from two different subnets with no direct connection.

> In either case, Cilium needs to run in encapsulation routing mode. However, when routing packets in above clusters, in many situations, we can get by native routing, and not pay the MTU overhead that comes with encapsulation. The reason for this inflexibility comes from the fact that Cilium datapath is not aware of the underlying network topology, and hence cannot make a decision between the two routing mode during runtime.

Author: Fixed

With higher adoption in cluster mesh and large scale clusters being common to support AI workloads, it would be beneficial to have a more performant dataplane that can optimize the throughput, while providing support for both encapsulation and native-routing.

## Goals

* Retain the behavior of current routing modes as-is
* Allow users to set a new routing mode called `hybrid`
* Allow users to pass in network topologies expressed through CIDR ranges during start-up

## Non-Goals

* Allow dynamic update to CIDR ranges (update without restart)

## Proposal

### Overview

> Add a new option for routing-mode config as follows `routing-mode: hybrid`. Also, add a new field in config to pass information about the CIDRs and connectivity. The CIDR information will persist in pinned eBPF maps so that routing decision can be made in runtime by eBPF programs.

Reviewer: Currently we have `routing-mode: native` + native routing CIDR that determines the directly reachable subnets (which implies that Cilium then masquerades to other destinations). I wonder if there's some conceptual overlap.

Author: Right, the idea is kind-of similar. In hybrid mode, we would allow native routing to ranges configured via option. Otherwise encapsulate/pass to Linux stack.

Reviewer: Similar or same?

I understand the idea that if we add another flag with its own parameters, that makes it easiest to avoid breaking some other existing functionality. At the same time, if the native routing CIDR determines the directly reachable subnet (and hence excludes traffic from SNAT/masquerade) and then we force the user to also separately configure the `subnet-topology`, then it seems like we're just asking the user to provide the same information twice in two different flags just because of some internal implementation details rather than because of the logical concept that's being configured.

Reviewer: I think this will be answered in other threads below from my latest review, so I'll resolve this thread.

### Examples

```

> subnet-topology: 10.0.0.1/24

Author: Agreed. Updating to reflect this.

```

> If source and destination IP belongs to above CIDR, route natively. Else, encapsulate the packet.

Reviewer: The "else encapsulate" seems to hint at a default route for tunneling traffic, is that intended? Typically today in tunnel mode, only traffic to known CIDRs (typically PodCIDRs) is encapsulated, then the remainder of the traffic is passed up to the stack to follow whatever default routing that Linux has configured.

Reviewer: I guess there's two notable points here: One is that "else encapsulate" doesn't currently make sense and would require some new notion of how to determine the remote tunnel endpoint for fallthrough traffic. The second point is that it is useful in many cases to be able to defer to the native Linux routing stack to determine forwarding behavior for traffic that Cilium isn't explicitly configured to handle.

> If source and destination IP belongs to above CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

Reviewer 4: Same for all occurrences.

```suggestion
If source and destination IP belongs to above CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to known CIDRs/pass to Linux stack).
```

```
subnet-topology: 10.0.0.1/24,10.10.0.1/24
```

The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
```
subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24
```

* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
* If source and destination IP belongs to `10.20.0.1/24`, route natively

> * Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

Author: Added a section to outline a datapath implementation.

## Impacts / Key Questions

### Impact: Network throughput

Given that today Cilium is encapsulating packets that could be natively routed, removing those scenarios would yield higher throughput. This is especially true in clusters without jumbo frames.

### Key Question: Datapath latency

In a highly fragmented network topography, lookup of IP to make routing decision may not be negligible.

## Future Milestones

### Deferred Milestone 1

Allow dynamic configuration of CIDRs (add/remove) without agent restart.

### Deferred Milestone 2

> Configure the subnets though a custom resource.

Author: Moved the dynamic update as a goal, I think it's best if user can add new CIDR block without restart.

```
subnet-topology: 10.0.0.1/24,10.10.0.1/24
```

The above two CIDR ranges have direct connectivity. If source and destination IP belongs to either CIDR, route natively. Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).
```
subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24
```

* `10.0.0.1/24` and `10.10.0.1/24` are connected and traffic can be natively routed
* If source and destination IP belongs to `10.20.0.1/24`, route natively
* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

### Control plane implementation

Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger dataplane reconcile.

### Datapath implementation

This section provides an overview of the datapath implementation. Consider the current implementation - as long as nodes have IP connectivity, Cilium can route traffic, both intra-cluster and inter-cluster (in case of cluster mesh). 

#### Case 1

> If `hybrid` mode is enabled without configuring any `subnet-topology`, routing mode will behave exactly same as encapsulation mode. Pod IPs can be from overlay or from a subnet, it will be encapsulated either way. So datapath implementation would be exactly as it is today for encapsulation routing mode.

Reviewer 5: Just on wording for clarity:
```suggestion
If `hybrid` mode is enabled without configuring any `subnet-topology`, it will behave the same as `tunnel` mode. Pod IPs can be from the overlay or from a subnet, traffic will be encapsulated either way. So datapath implementation would be exactly as it is today for the `tunnel` routing mode.
```

#### Case 2

If `subnet-topology` is configured, the CIDR ranges can be configured as part of node configs, similar to [IPv4PodSubnets](https://github.com/cilium/cilium/blob/f78aca7b5dff52e6d07723f6577dd0d3f913fea6/pkg/datapath/types/node.go#L177).

Similar to how masquerade uses `NativeRoutingCIDRIPv4` to check if [SNAT is needed](https://github.com/cilium/cilium/blob/f78aca7b5dff52e6d07723f6577dd0d3f913fea6/bpf/lib/nat.h#L713), this implementation would add a similar function that can check source and destination address against multiple CIDRs. However, in this case, since we will be storing multiple CIDRs, we will use an eBPF map of type `BPF_MAP_TYPE_LPM_TRIE` to efficiently search the map. 

Next, while making routing decision today, in both `lxc` and `host` eBPF programs, a check is made to see if tunneling should be done. This implementation would add an extra check to tunneling decision. The check would use above function to figure out if native routing can be done. 
- If yes, we skip tunneling and the packet is instead sent to Linux stack for handling, where the underlying networking would route the packet to destination. 

> - If no, tunneling happens exactly as it works today 

Reviewer 5: Could you describe how this check would be implemented exactly? In particular to handle the case `subnet-topology: 10.0.0.1/24,10.10.0.1/24;10.20.0.1/24` described above. I don't think it's as simple as skipping tunneling if both the src and dst IP addresses are found in the LPM map.

**Authors:** Anubhab Majumdar <anmajumdar@microsoft.com>, Vamsi Kalapala <vakr@microsoft.com>, Krunal Jain <krunaljain@microsoft.com>

> **Status:**

Reviewer 2: Please add a status here

https://github.com/cilium/design-cfps?tab=readme-ov-file#status

Author: Adding `Implementable` here.

* If source and destination IP belongs to `10.20.0.1/24`, route natively
* Else, preserve current behavior (encapsulate if being routed to know CIDRs/pass to Linux stack).

### Control plane implementation

> Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger dataplane reconcile.

Reviewer 5: Having read the datapath implementation section, I don't understand why a full datapath reconcile would be needed. Can't we simply update the LPM map?

> Similar to dynamic flowlogs or Hubble dynamic exporter, allow users to configure the CIDRs on the fly through a configmap. The configmap will be mounted on each agent. On create/update, update `LocalNodeConfig` and trigger dataplane reconcile.

Author: I should have been explicit about this. I meant reconcile the config to the eBPF map. This shouldn't need full datapath restart that disrupts networking. I will rewrite this section to clarify my point.
